EquiTrack/partners/tests/test_api_interventions.py

```
@@ -614,3 +615,145 @@ def test_group_permission_non_staff(self):
+        # The id of the newly-created lower result should be associated with my result link, and it should be
+        # the only one associated with that result link.
+        self.assertEqual([response_json['id']],
+                         [lower_result.id for lower_result in self.result_link.ll_results.all()])
+        self.assertEqual(response_json.get('name'), 'my lower result')
+
+        return response_json
+
+    def test_no_permission_user_forbidden(self):
+        '''Ensure a non-staff user gets the 403 smackdown'''
+        response = self._make_request(UserFactory())
+        self.assertEquals(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_unauthenticated_user_forbidden(self):
+        '''Ensure an unauthenticated user gets the 403 smackdown'''
+        factory = APIRequestFactory()
+        view_info = resolve(self.url)
+        request = factory.post(self.url, data=self.data, format='json')
+        response = view_info.func(request)
+        self.assertEquals(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_group_permission_non_staff(self):
+        '''Ensure group membership is sufficient for create; even non-staff group members can create'''
+        user = UserFactory()
+        response = self._make_request(user)
+        self.assertEquals(response.status_code, status.HTTP_403_FORBIDDEN)
+
+        _add_user_to_partnership_manager_group(user)
+
+        # Now the request should succeed.
+        response = self._make_request(user)
+        self.assertResponseFundamentals(response)
+
+    def test_code_read_only(self):
+        '''Ensure lower_result.code can't be written'''
+        user = UserFactory()
+        _add_user_to_partnership_manager_group(user)
+        data = self.data.copy()
+        data['code'] = 'ZZZ'
+
+        response = self.forced_auth_req('post', self.url, user=user, data=data)
+        response_json = self.assertResponseFundamentals(response)
```

not really a concern with this PR but more just a question on my side - is it the case that DRF just silently swallows invalid params on all create / update requests?

---

Great question! Some of DRF's behavior surprises me. I expected this request to give a 400 when I first tried it.

The exact answer depends on the definition of "invalid". (I had to look at DRF's serializer code to figure that out. It was educational!) The reason that DRF silently ignores this attempt to set `code` is because that field is defined as read-only on the serializer. If a serializer gets incoming data, it knows it's handling a write request (create/update), so it only concerns itself with writeable fields. Data associated with non-writeable fields is dropped. That's here --
https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L455

That means you can also cram irrelevant fields into the data dict and populate them with lorem ipsum and DRF won't bat an eyelash. However, if you write bad data to a writeable field (e.g. 'lorem ipsum' where it's expecting a PK), then validation will fail, ValidationError will be raised, and you'll get a 400.

---

nice, thanks for the comprehensive explanation!

---

yw. As I said, this concept was fuzzy for me and it was nice to get it straight. I hope that typing it out will help me remember it.